src/lib/dovetail_audio.ts

```diff
@@ -0,0 +1,221 @@
+
+type AllUnion = [
+  DovetailResponse | PromiseLike<DovetailResponse>,
+  AdzerkResponse | PromiseLike<AdzerkResponse>
+];
+
+type AllResult = [DovetailResponse, AdzerkResponse];
+
+interface Arrangement {
+  entries: ArrangementEntry[];
+  duration?: number;
+}
+
+function sumDuration(collector: number, entry: {duration?: number}) {
+  return collector + entry.duration || 0;
+}
+
+const DURATION_CHANGE = 'durationchange';
+const TIME_UPDATE = 'timeupdate';
+const ENDED = 'ended';
+const AD_START = 'adstart';
+const AD_END = 'adend';
+
+class DovetailEvent {
+  static build(eventName: string, audio: DovetailAudio, extras?: {}) {
+    let event = new Event(eventName);
+    Object.defineProperty(event, 'target', {
+      value: audio,
+      writable: false
+    });
+    Object.defineProperty(event, 'currentTarget', {
+      value: audio,
+      writable: false
+    });
+    if (extras) {
+      for (let property in extras) {
+        if (extras.hasOwnProperty(property)) {
+          event[property] = extras[property];
+        }
+      }
+    }
+    return event;
+  }
+}
+
+export class DovetailAudio extends ExtendableAudio {
+  private arrangement: Arrangement = {entries: []};
+  private index: number = 0;
+
+  private _dovetailLoading = false;
+
+  private adzerkFetcher = new AdzerkFetcher();
+  private dovetailFetcher = new DovetailFetcher();
+
+  private currentPromise: Promise<any>;
+  private currentAdzerkPromise: Promise<AdzerkResponse>;
+  private resumeOnLoad = false;
+
+  constructor(url: string) {
+    super(url);
+    this._audio.addEventListener(DURATION_CHANGE, this.listenerOnDurationChange.bind(this));
+    this._audio.addEventListener(ENDED, this.listenerOnEnded.bind(this));
+    this._audio.addEventListener(TIME_UPDATE, this.listenerOnTimeUpdate.bind(this));
+    this.finishConstructor();
+  }
+
+  play() {
+    if (this._dovetailLoading) {
+      this.resumeOnLoad = true;
+    } else {
+      this._audio.play();
+    }
+  }
+
+  pause() {
+    if (this._dovetailLoading) {
+      this.resumeOnLoad = false;
+    } else {
+      this._audio.pause();
+    }
+  }
+
+  get duration() {
+    if (typeof this.arrangement.duration === 'undefined') {
+      this.arrangement.duration = this.arrangement.entries.reduce(sumDuration, 0);
+    }
+    return this.arrangement.duration;
+  }
+
+  get currentTime() {
+    return this._audio.currentTime +
+      this.arrangement.entries.slice(0, this.index).reduce(sumDuration, 0);
+  }
+
+  set currentTime(position: number) {
+    if (this.duration >= position) {
+      let soFar = 0, paused = this.paused;
+      for (let i = 0; i < this.arrangement.entries.length; i++) {
+        let duration = this.arrangement.entries[i].duration;
+        if (soFar + duration > position) {
+          this.skipToFile(i);
+          this._audio.currentTime = position - soFar;
+          if (!paused) { this.play(); }
+          return;
+        } else {
+          soFar += duration;
+        }
+      }
+    }
+  }
```

Comment:
pretty basic implementation of seeking, we walk through the arrangement until we find the file we want and then seek to the right spot within it.